core/src/main/java/org/apache/calcite/plan/volcano/VolcanoRuleCall.java

```diff
@@ -285,11 +288,38 @@ private void matchRecurse(int solve) {
         final int parentOrdinal = operand.getParent().ordinalInRule;
         final RelNode parentRel = rels[parentOrdinal];
         final List<RelNode> inputs = parentRel.getInputs();
-        if (operand.ordinalInParent < inputs.size()) {
+        // if the child is unordered, then add all rels in all input subsets to the successors list
+        // because unordered can match child in any ordinal
+        if (parentOperand.childPolicy == RelOptRuleOperandChildPolicy.UNORDERED) {
+          if (operand.getMatchedClass() == RelSubset.class) {
+            successors = inputs;
+          } else {
+            List<RelNode> allRelsInAllSubsets = new ArrayList<>();
+            Set<RelNode> duplicates = new HashSet<>();
+            for (RelNode input : inputs) {
+              if (!duplicates.add(input)) {
+                // Ignore duplicate subsets
+                continue;
+              }
+              RelSubset inputSubset = (RelSubset) input;
+              for (RelNode rel : inputSubset.getRels()) {
+                if (!duplicates.add(rel)) {
+                  // Ignore duplicate relations
+                  continue;
+                }
+                allRelsInAllSubsets.add(rel);
+              }
+            }
+            successors = allRelsInAllSubsets;
+          }
+        } else if (operand.ordinalInParent < inputs.size()) {
+          // child policy is not unordered
+          // we need to find the exact input node based on child operand's ordinalInParent
           final RelSubset subset =
               (RelSubset) inputs.get(operand.ordinalInParent);
           if (operand.getMatchedClass() == RelSubset.class) {
-            successors = subset.set.subsets;
```

Comment:
@julianhyde , I don't think `operand(RelSubset.class)` was often used in practice, however I don't think `subset.set.subsets` is a valid `successors`.
What it does, it provides all the subsets of a set, no matter which was the actual input for the relation.

I guess the only viable solution here is `ImmutableList.of(subset)`. In other words, it should provide the subset itself rather than trying to iterate over relations in the subset.

See relevant test correction `VolcanoPlannerTest.java`:  the rule no longer fires for `"PhysSingleRel:Subset#0.NONE"` which was impossible (PhysSingleRel always conumes `.PHYS` subset, and it makes no sense to fire a rule for non-existing combination of `Phys` top level relation and `NONE` downstream subset.

Reply:
> I don't think subset.set.subsets is a valid successors. [...] I guess the only viable solution here is ImmutableList.of(subset).

I agree.